R/utils-multi.R

```diff
@@ -0,0 +1,79 @@
+multi_dots <- function(...,
+                       .multi = c("error", "comma", "pipe", "explode"),
+                       error_arg = "...",
+                       error_call = caller_env()) {
+  if (is.function(.multi)) {
+    check_function2(.multi, call = error_call, arg = ".multi")
+  } else {
+    .multi <- arg_match(.multi, error_arg = ".multi", error_call = error_call)
+  }
+
+  dots <- list2(...)
+  if (length(dots) == 0) {
+    return(list())
+  }
+
+  if (!is_named(dots)) {
+    cli::cli_abort(
+      "All components of {.arg {error_arg}} must be named.",
+      call = error_call
+    )
+  }
+
+  type_ok <- map_lgl(dots, function(x) is_atomic(x) || is.null(x))
+  if (any(!type_ok)) {
+    cli::cli_abort(
+      "All elements of {.arg {error_arg}} must be either an atomic vector or NULL.",
+      call = error_call
+    )
+  }
+
+  n <- lengths(dots)
+  if (any(n > 1)) {
+    if (is.function(.multi)) {
+      dots[n > 1] <- lapply(dots[n > 1], format_query_param)
+      dots[n > 1] <- lapply(dots[n > 1], .multi)
+      dots[n > 1] <- lapply(dots[n > 1], I)
+    } else if (.multi == "comma") {
+      dots[n > 1] <- lapply(dots[n > 1], format_query_param)
+      dots[n > 1] <- lapply(dots[n > 1], paste0, collapse = ",")
+      dots[n > 1] <- lapply(dots[n > 1], I)
+    } else if (.multi == "pipe") {
+      dots[n > 1] <- lapply(dots[n > 1], format_query_param)
+      dots[n > 1] <- lapply(dots[n > 1], paste0, collapse = "|")
+      dots[n > 1] <- lapply(dots[n > 1], I)
+    } else if (.multi == "explode") {
+      dots <- explode(dots)
+      dots[n > 1] <- lapply(dots[n > 1], format_query_param)
+      dots[n > 1] <- lapply(dots[n > 1], I)
+    } else if (.multi == "error") {
+      cli::cli_abort(
+        c(
+          "All vector elements of {.arg {error_arg}} must be length 1.",
+          i = "Use {.arg .multi} to choose a strategy for handling vectors."
+        ),
+        call = error_call
+      )
+    }
+  }
+
+  # Format other params
+  dots[n == 1] <- lapply(dots[n == 1], format_query_param, error_call = error_call)
```

Comment:
I stumbled over this issue

``` r
devtools::load_all("~/GitHub/httr2/")
#> ℹ Loading httr2
data <- list(a = c("ab", "c"))

request_test("/post") %>%
  req_body_form(!!!data, .multi = "comma") %>%
  req_verbose(body_req = TRUE) %>% 
  req_perform()
#> -> POST /post HTTP/1.1
#> -> Host: 127.0.0.1:50747
#> -> User-Agent: httr2/0.2.3.9000 r-curl/5.1.0 libcurl/8.1.2
#> -> Accept: */*
#> -> Accept-Encoding: deflate, gzip
#> -> Content-Type: application/x-www-form-urlencoded
#> -> Content-Length: 9
#> -> 
#> >> a=ab,c%20
#> <- HTTP/1.1 200 OK
#> <- Connection: close
#> <- Date: Fri, 10 Nov 2023 10:00:31 GMT
#> <- Content-Type: application/json
#> <- Content-Length: 457
#> <- ETag: "6dd433e5"
#> <-
#> <httr2_response>
#> POST http://127.0.0.1:50747/post
#> Status: 200 OK
#> Content-Type: application/json
#> Body: In memory (457 bytes)
```

<sup>Created on 2023-11-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

You can see that the body is `a=ab,c%20` instead of `a=ab,c` because white spaces are appended.
In `format_query_param()` we need to use `justify = "none"`

Reply:
Hmmm, I did that in #355. Maybe that didn't get in this branch? Anyway I'll double check.

Reply:
No looks like a different problem.

Reply:
ARGH you need `justify = "none"` for strings and `trim = TRUE` for numbers.